### PR TITLE
*: switch to pnpm

### PR DESCRIPTION
**Pull request by sjbarag**
# * // switch to pnpm

We're moving in this direction in other repos due to Bazel's `rules_js` — may as well be consistent.

#### Checklist
- [x] I have written or updated test for the changes I made
- [x] I have updated the README of the package I'm working in to reflect my changes
- [x] I have added or updated Storybook if appropriate for my changes

### CODE REVIEW

File: packages/design-tokens/package.json [hunk 1]
```diff
@@ -13,9 +13,7 @@
   "types": "dist/web/tokens.d.ts",
   "scripts": {
     "build": "style-dictionary build",
-    "clean": "style-dictionary clean",
-    "prepare": "npm run clean && npm run build",
```
Comment (sjbarag) [inline packages/design-tokens/package.json:17]: Something about the "automatically build after running `pnpm install`" behavior felt odd, so I removed it 😄

File: packages/icons/package.json [hunk 2]
```diff
@@ -8,15 +8,14 @@
   "main": "dist/index.js",
   "types": "dist/index.d.ts",
   "scripts": {
-    "build": "npm run build:generate:stripfill && npm run build:generate:preserve && npm run build:typescript",
+    "build": "run-s build:generate:stripfill build:generate:preserve build:typescript",
     "build:typescript": "tsc",
     "build:generate:stripfill": "svgr --typescript --icon -d src/components svg/customFill",
     "build:generate:preserve": "svgr --typescript --icon --no-runtime-config -d src/components svg/preserveFill",
-    "build:watch": "tsc --watch",
+    "build:watch": "tsc --watch --preserveWatchOutput",
```
Comment (sjbarag) [inline packages/icons/package.json:15]: Without this, `tsc --watch` clears the terminal while other watch tasks don't and it looks weird!

File: packages/storybook-ui-components/package.json [hunk 3]
```diff
@@ -18,8 +18,8 @@
   },
   "homepage": "https://github.com/cockroachdb/ui#readme",
   "dependencies": {
-    "@cockroachlabs/icons": "^0.7.6",
-    "@cockroachlabs/ui-components": "^0.7.7",
+    "@cockroachlabs/icons": "workspace:../icons",
+    "@cockroachlabs/ui-components": "workspace:../ui-components",
```
Comment (sjbarag) [inline packages/storybook-ui-components/package.json:22]: hooray `workspace:` protocol: https://pnpm.io/workspaces#workspace-protocol-workspace 🎉

File: packages/ui-components/package.json [hunk 4]
```diff
@@ -9,27 +9,27 @@
...
+    "start": "pnpm run build:watch"
   },
   "keywords": [],
   "license": "MIT",
   "dependencies": {
-    "@cockroachlabs/icons": "^0.7.6",
+    "@cockroachlabs/icons": "workspace:../icons",
     "@popperjs/core": "^2.9.2",
-    "@types/recharts": "^1.8.23",
+    "lodash": "^4.17.21",
```
Comment (sjbarag) [inline packages/ui-components/package.json:32]: Accidentally-missing dependency (previously hoisted from another package)

File: packages/ui-components/package.json [hunk 5]
```diff
@@ -9,27 +9,27 @@
...
-    "start": "npm run build:watch"
+    "start": "pnpm run build:watch"
   },
   "keywords": [],
   "license": "MIT",
   "dependencies": {
-    "@cockroachlabs/icons": "^0.7.6",
+    "@cockroachlabs/icons": "workspace:../icons",
     "@popperjs/core": "^2.9.2",
-    "@types/recharts": "^1.8.23",
```
Comment (sjbarag) [inline packages/ui-components/package.json:32]: This was never a runtime dependency — moved to dev dependencies

File: packages/ui-components/package.json [hunk 6]
```diff
@@ -52,6 +52,8 @@
     "@types/node": "16.18.14",
     "@types/react": "17.0.53",
     "@types/react-lines-ellipsis": "0.15.1",
+    "@types/recharts": "^1.8.23",
+    "@types/testing-library__jest-dom": "^5.14.6",
```
Comment (sjbarag) [inline packages/ui-components/package.json:56]: Accidentally-missing dependency (previously provided by `@testing-library/jest-dom` as a transitive dependency)

File: packages/ui-components/package.json [hunk 7]
```diff
@@ -68,6 +70,7 @@
     "jest-environment-jsdom": "29.4.3",
     "jest-runner-eslint": "1.1.0",
     "mini-css-extract-plugin": "^2.7.5",
+    "npm-run-all": "4.1.5",
```
Comment (sjbarag) [inline packages/ui-components/package.json:73]: Accidentally missing dependency (got hoisted from another package)

File: packages/ui-components/webpack.config.js [hunk 8]
```diff
@@ -1,7 +1,6 @@
 /* eslint-disable @typescript-eslint/no-var-requires */
 /* globals module, __dirname */
 const path = require("path");
-const WebpackBar = require("webpackbar");
```
Comment (sjbarag) [inline packages/ui-components/webpack.config.js:4]: It got in the way of watch task reporting and spewed way too many lines. I can put it back if we want, but 🤷

**Review by nathanstilwell — COMMENTED** [review]
So this review looks good, but there's one big catch, and that is that I can't get `auto` publishing to work with this setup. Here's my [wip branch](https://github.com/cockroachdb/ui/tree/nathanstilwell/switch-to-pnpm-wipv) for reference.

So `auto` [depends on `lerna`](https://github.com/intuit/auto/issues/1698#issuecomment-760573893) to determine what changed and needs to publish. I tried adding `lerna` back and configuring for `pnpm` (that part might work correctly). But when I run `auto` I get an error in `pnpm`,

```shell
╔═══════════════════════════
║ ui on  switch-to-pnpm [⇡] via  v16.20.1
╚════  △ pnpm run release

> root@ release /Users/nathanstilwell/code/ui
> auto shipit

⚠  warning   lerna notice cli v7.1.1
lerna success found 5 packages

TypeError: Cannot read properties of undefined (reading 'reduce')
    at findPackages (/Users/nathanstilwell/code/ui/node_modules/.pnpm/get-monorepo-packages@1.2.0/node_modules/get-monorepo-packages/index.js:17:6)
    at Object.getPackages [as default] (/Users/nathanstilwell/code/ui/node_modules/.pnpm/get-monorepo-packages@1.2.0/node_modules/get-monorepo-packages/index.js:37:14)
    at getMonorepoPackage (/Users/nathanstilwell/code/ui/node_modules/.pnpm/@auto-it+npm@10.43.0_@types+node@20.4.0_typescript@5.1.6/node_modules/@auto-it/npm/src/index.ts:158:31)
    at /Users/nathanstilwell/code/ui/node_modules/.pnpm/@auto-it+npm@10.43.0_@types+node@20.4.0_typescript@5.1.6/node_modules/@auto-it/npm/src/index.ts:1098:30
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
 ELIFECYCLE  Command failed with exit code 1.
```

I did not dig in very far here. In my mind we have a couple of options to proceed forward,

1) Debug this error and submit a patch for `pnpm` or `auto` or whatever to resolve the issue.
2) Create our own [plugin](https://intuit.github.io/auto/docs/plugins/writing-publishing-plugins) for `auto`.
3) Drop `auto` entirely and create a solution using only pnpm and Github Actions.

Happy to brainstorm further.

File: packages/storybook-ui-components/package.json (hunk 3 above)
Comment (sjbarag) [reply packages/storybook-ui-components/package.json:22]: It shouuuuuuuld automatically link everything together :)

File: packages/icons/package.json [hunk 9]
```diff
@@ -8,15 +8,14 @@
   "main": "dist/index.js",
   "types": "dist/index.d.ts",
   "scripts": {
-    "build": "npm run build:generate:stripfill && npm run build:generate:preserve && npm run build:typescript",
+    "build": "run-s build:generate:stripfill build:generate:preserve build:typescript",
```
[the remark this answers is not in the document]
Comment (sjbarag) [reply packages/icons/package.json:11]: Unfortunately it's an alias that comes wioth `npm-run-all`. I don't think `pnpm` supports sequential script targets natively :thinking:

**Comment by sjbarag** [comment]
I _thiiiink_ I've got `auto` working now! Had to upgrade to 11.x and lerna 7.x, but everything seems good. Note that `auto shipit --dry-run` doesn't work due to https://github.com/intuit/auto/issues/2129. Something's just weird with `--dry-run` 🤷

**Review by nathanstilwell — APPROVED** [review]
I guess we just had to wait until Lerna and pnpm learned to play together. 😄

**Pull request merged**